common/lib/api-client/middleware/process-response.test.js

```diff
@@ -0,0 +1,139 @@
+const proxyquire = require('proxyquire')
+
+const findUnpopulatedResourcesStub = sinon.stub()
+
+const middleware = proxyquire('./process-response', {
+  '../../find-unpopulated-resources': findUnpopulatedResourcesStub,
+})
+
+describe('API Client', function () {
+  describe('Response processing middleware', function () {
+    describe('#process-response', function () {
+      context('when payload includes a response', function () {
+        it('should return payload as is', function () {
+          const payload = { req: {}, res: {} }
+          expect(middleware.req(payload)).to.deep.equal({ ...payload })
+        })
+      })
+
+      context('when payload contains a response', function () {
+        let response
+        let processedResponse
+
+        beforeEach(function () {
+          findUnpopulatedResourcesStub.resetHistory()
+          response = {
+            data: {
+              data: {},
+              included: [{ id: 'foo', type: 'foos' }],
+            },
+            req: {},
+          }
+        })
+
+        context('and the response should not be processed', function () {
+          beforeEach(function () {
+            processedResponse = middleware.req(
+              JSON.parse(JSON.stringify(response))
+            )
+          })
+          it('should return the response unchanged', function () {
+            expect(processedResponse).to.deep.equal(response)
+          })
+        })
+
+        context('and the response should be processed', function () {
+          beforeEach(function () {
+            findUnpopulatedResourcesStub.returns([
+              {
+                id: 'bar',
+                type: 'bars',
+              },
+            ])
+            response.req.preserveResourceRefs = true
```

Comment:
This line is shared by all the child contexts, not just `when preserveResourceRefs is set to true` so that's slightly confusing

Reply:
Yes. It's meant to be. We do this all over the place rather than setting stuff up explicitly.

This pattern (and it is confusing at times) is one of the reasons I don't like Mocha.

Reply:
Sorry, I was unclear - what I meant the context below refers to "when preservedResourceRefs is true", but then there are other contexts being that are using this "is true" as well, so the ordering is a little confusing.

It's not anything major :)